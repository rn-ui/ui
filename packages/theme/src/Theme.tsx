/* tslint:disable */
import deepMerge from './tools/deepMerge';
import mergeComponentAndThemes from './tools/mergeComponentAndThemes';
import { resolveIncludes } from './tools/resolveIncludes';
import normalizeTheme from './normalizer/normalizeTheme';
import defaultTheme from './resources/default';

interface Themes {
  default: object;
  light: object;
  dark: object;
}

// @ts-ignore
const themes: Themes = {};

// Privates, ideally those should be symbols
const THEME_STYLES = '@southem.theme';

const resolveStyle = (style, baseStyle) => normalizeTheme(resolveIncludes(style, baseStyle));

/**
 * Theme
 *
 * @example
 * const defaultTheme = {};
 * const defaultThemePro = {};
 *
 * Theme.registerTheme('default', [
 *   defaultTheme,
 *   defaultThemePro,
 * ]);
 */
class Theme {

  constructor(themeStyle) {
    // @ts-ignore
    // this[THEME_STYLES] = resolveStyle(themeStyle);
  }

  /**
   * registerTheme
   * @param {String} themeName
   * @param {Array|Object} styleSheets
   */
  static registerTheme(themeName, styleSheets: Array<Object>|Object): void {
    let mergedStyleSheets = styleSheets;

    if (Array.isArray(styleSheets)) {
      mergedStyleSheets = deepMerge({}, ...styleSheets);
    }

    // themes[themeName] = new Theme(themeName, mergedStyleSheets);
    // @ts-ignore
    themes[themeName] = resolveStyle(mergedStyleSheets);
  }

  /**
   * registerDefaultTheme
   * @param {Object} styleSheets
   */
  static registerDefaultTheme(styleSheets): void {
    Theme.registerTheme('default', styleSheets);
  }

  /**
   * getTheme
   * @param {String} themeName
   * @return {Object}
   */
  static getTheme(themeName): Object {
    // @ts-ignore
    if (themeName === 'default' && !themes.default) {
      Theme.registerDefaultTheme(defaultTheme);
    }

    return themes[themeName];
  }

  /**
   * getDefaultTheme
   * @return {Object}
   */
  static getDefaultTheme(): Object {
    return Theme.getTheme('default');
  }

  /**
   * isDefine
   * @param {String} themeName
   * @return {Boolean}
   */
  static isDefine(themeName): Boolean {
    return !!themes[themeName];
  }
}

export {
  Theme as default,
  themes,
};
