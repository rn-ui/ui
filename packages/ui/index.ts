/* tslint:disable */
export * from './src/tools';
export * from './src/common';
export * from './src/resources';
// export * from './ui';
