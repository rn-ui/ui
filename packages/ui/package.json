{
  "name": "@southem/ui",
  "description": "Customizable set of components for React Native applications",
  "version": "0.0.1",
  "main": "index.js",
  "author": "Southem <contact@southem.com>",
  "license": "MIT",
  "homepage": "https://github.com/southem/southem#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/southem/southem.git"
  },
  "bugs": {
    "url": "https://github.com/southem/southem/issues"
  },
  "keywords": [
    "react-native",
    "southem",
    "kit",
    "ui"
  ],
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "lodash": "^4.6.1",
    "color": "^3.1.2"
  },
  "peerDependencies": {
    "@southem/theme": "^0.0.1",
    "@southem/animation": "^0.0.1",
    "react-native-svg": "^9.4.0"
  },
  "rnpm": {
    "assets": [
      "assets/fonts"
    ],
    "commands": {
      "postlink": "node ./node_modules/@southem/ui/scripts/link.js"
    }
  }
}
