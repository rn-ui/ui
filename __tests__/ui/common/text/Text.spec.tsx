import React from 'react';
import {
  Text as RNText,
  StyleSheet,
} from 'react-native';
import {
  render,
  shallow,
  fireEvent,
  RenderAPI,
} from 'react-native-testing-library';
import { ReactTestInstance } from 'react-test-renderer';
import Theme, { ThemeProvider } from '@southem/theme';
import {
  darkTheme,
  Text,
  TextProps,
} from '@southem/ui';

import { typography } from '@southem/ui';

Theme.registerDefaultTheme(darkTheme);

const Mock = (props?: TextProps): React.ReactElement<{}> => {
  return (
    <ThemeProvider
      theme={'default'}>
      <Text {...props} />
    </ThemeProvider>
  );
};

const renderComponent = (props?: TextProps): RenderAPI => {
  return render(
    <Mock {...props} />,
  );
};

describe('@text: matches snapshot', () => {
  describe('* interaction', () => {
    it('* stateless', () => {
      const component: RenderAPI = renderComponent();
      const { output } = shallow(component.getByType(Text));

      expect(output).toMatchSnapshot();
    });
  });
  describe('* appearance', () => {

    const text: React.ReactText = 'TEXT';

    it('* empty', () => {
      const component: RenderAPI = renderComponent();
      const { output } = shallow(component.getByType(Text));

      expect(output).toMatchSnapshot();
    });

    it('* text', () => {
      const component: RenderAPI = renderComponent({ children: text });
      const { output } = shallow(component.getByType(Text));

      expect(output).toMatchSnapshot();
    });

    it('* text * category h1 and font size of 50', () => {
      const component: RenderAPI = renderComponent({
        category: 'h1',
        children: text,
      });
      const componentText: ReactTestInstance = component.getByType(RNText);

      const styles  = StyleSheet.flatten(componentText.props.style);
      const { output } = shallow(componentText);

      expect(styles.fontFamily).toEqual(typography.regular.fontFamily);
      expect(styles.fontSize).toBe(50);
      expect(output).toMatchSnapshot();
    });

    it('* custom text (styled)', () => {
      const component: RenderAPI = renderComponent({
        children: text,
        style: [{ color: 'red' }, { fontSize: 30 }],
      });
      const componentText: ReactTestInstance = component.getByType(RNText);

      const styles  = StyleSheet.flatten(componentText.props.style);
      const { output } = shallow(componentText);

      expect(styles).toEqual({
        ...typography.regular,
        color: 'red',
        fontSize: 30,
      });

      expect(output).toMatchSnapshot();
    });
  });

});
