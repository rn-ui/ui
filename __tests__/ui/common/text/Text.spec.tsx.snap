// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@text: matches snapshot * appearance * category h1 and text 1`] = `
<Text
  animationOptions={Object {}}
  onLayout={[Function]}
  style={
    Object {
      "color": "rgba(0, 0, 0, 0.5)",
      "fontFamily": "Fira Sans",
      "fontSize": 50,
      "fontWeight": "700",
    }
  }
>
  TEXT
</Text>
`;

exports[`@text: matches snapshot * appearance * custom text (styled) 1`] = `
<Text
  animationOptions={Object {}}
  onLayout={[Function]}
  style={
    Object {
      "color": "red",
      "fontFamily": "Fira Sans",
      "fontSize": 30,
      "fontWeight": "400",
    }
  }
>
  TEXT
</Text>
`;

exports[`@text: matches snapshot * appearance * empty 1`] = `
<@theme(Animated(Text))
  animationOptions={Object {}}
  forwardedRef={null}
/>
`;

exports[`@text: matches snapshot * appearance * text * category h1 and font size of 50 1`] = `
<Text
  animationOptions={Object {}}
  onLayout={[Function]}
  style={
    Object {
      "color": "rgba(0, 0, 0, 0.5)",
      "fontFamily": "Fira Sans",
      "fontSize": 50,
      "fontWeight": "700",
    }
  }
>
  TEXT
</Text>
`;

exports[`@text: matches snapshot * appearance * text 1`] = `
<@theme(Animated(Text))
  animationOptions={Object {}}
  forwardedRef={null}
>
  TEXT
</@theme(Animated(Text))>
`;

exports[`@text: matches snapshot * interaction * stateless 1`] = `
<@theme(Animated(Text))
  animationOptions={Object {}}
  forwardedRef={null}
/>
`;
